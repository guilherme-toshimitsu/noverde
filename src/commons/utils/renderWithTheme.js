import React from 'react';
import { ThemeProvider } from 'styled-components';
import defaultTheme from '../theme';
import { render } from '@testing-library/react';

function renderWithTheme(children) {
  return render(<ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>);
}

export { renderWithTheme };
