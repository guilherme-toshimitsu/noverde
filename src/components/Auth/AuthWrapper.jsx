import React, { useReducer } from 'react';
import PropTypes from 'prop-types';

import { createReducer } from '@utils/createReducer';
// import Api from '@api';
// import { useSnackBar } from '@components/SnackBar';

import * as actionTypes from './Reducer/authActionTypes';
import actions from './Reducer/authActions';
import initialState from './Reducer/authState';
import { AuthContextProvider } from './authContext';

const reducer = createReducer(actions);

const AuthWrapper = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const doLogin = (cpf) => {
    dispatch({ type: actionTypes.LOGIN, cpf });
  };

  const doLogout = () => {
    dispatch({ type: actionTypes.LOGOUT });
  };

  const actionResolver = {
    doLogin: doLogin,
    doLogout: doLogout,
  };
  return (
    <AuthContextProvider value={[state, actionResolver]}>
      {children}
    </AuthContextProvider>
  );
};

AuthWrapper.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]).isRequired,
};

export default AuthWrapper;
