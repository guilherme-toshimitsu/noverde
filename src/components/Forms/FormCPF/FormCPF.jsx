import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import Button from '@components/Button';
import {
  StyledInput,
  StyledP,
  StyledInputContainer,
  StyledTitle,
} from '@components/Input';

import { isValidCpf } from '@commons/utils/validations/cpf';
import useDebounce from '@commons/utils/debounce';

import { useAuth } from '@components/Auth';

const FormCPF = ({ onSuccess }) => {
  const [CPF, setCPF] = useState(null);
  const [error, setError] = useState(false);
  const [{}, { doLogin }] = useAuth();
  const [submit, setSubmit] = useState(false);
  const debouncedButton = useDebounce(submit, 500);

  useEffect(() => {
    setError(CPF && isValidCpf(CPF));
  }, [CPF]);

  useEffect(() => {
    if (debouncedButton) {
      doLogin(CPF);
      onSuccess();
    }
  }, [debouncedButton]);

  const onSubmit = () => {
    setSubmit(true);
  };

  return (
    <>
      <StyledTitle>
        {'Ola, queremos te ver #NOVERDE, para isso precisamos do seu CPF'}
      </StyledTitle>
      <StyledInputContainer>
        <StyledInput
          data-testid="input-cpf"
          placeholder={'Informe o seu CPF'}
          onBlur={(event) => setCPF(event.target.value)}
          error={error}
          onKeyDown={(e) => {
            e.keyCode === 13 ? e.target.blur() : null;
          }}
        />
        {error && <StyledP data-testid="error-cpf">{error}</StyledP>}
      </StyledInputContainer>
      <Button
        disabled={!CPF || !!error}
        data-testid={'button-cpf'}
        onClick={() => onSubmit()}
        color="primary"
        variant="contained"
      >
        {'PROXIMO'}
      </Button>
    </>
  );
};

FormCPF.propTypes = {
  onSuccess: PropTypes.func,
};

FormCPF.defaultProps = {
  onSuccess: () => {},
};

export default FormCPF;
