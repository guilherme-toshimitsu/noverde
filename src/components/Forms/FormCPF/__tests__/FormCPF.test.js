import React from 'react';
import { fireEvent, wait } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import AuthWrapper from '@components/Auth';

import FormCPF from '../';
import { renderWithTheme } from '@commons/utils/renderWithTheme'
describe('<FormCPF> loads', () => {
  it('renders', () => {
    const container = renderWithTheme(
    <>
      <AuthWrapper>
        <BrowserRouter>
          <FormCPF />
        </BrowserRouter>
      </AuthWrapper>
    </>
  );
    expect(container.asFragment()).toMatchSnapshot();
  });

  //Text invalid message on failed InputCPF
  it('input a failed value', async () => {
    const container = renderWithTheme(
    <>
      <AuthWrapper>
        <BrowserRouter>
          <FormCPF />
        </BrowserRouter>
      </AuthWrapper>
    </>
  );

    const input = container.getByTestId('input-cpf')
    await wait(() => fireEvent.change(input, { target: { value: '40000' }}))
    await wait(() => fireEvent.blur(input))

    await wait(() => container.getByTestId('error-cpf'))
    const error = container.getByTestId('error-cpf')

    expect(error).toMatchSnapshot();
  });

});
