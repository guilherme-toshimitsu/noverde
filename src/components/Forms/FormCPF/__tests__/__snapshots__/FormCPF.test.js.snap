// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormCPF> loads input a failed value 1`] = `
<p
  class="sc-gKAblj bMOaAl"
  data-testid="error-cpf"
>
  CPF Invalido
</p>
`;

exports[`<FormCPF> loads renders 1`] = `
<DocumentFragment>
  <h2
    class="sc-jSFkmK bUgUTJ"
  >
    Ola, queremos te ver #NOVERDE, para isso precisamos do seu CPF
  </h2>
  <div
    class="sc-gtssRu jQJSzM"
  >
    <input
      class="sc-eCApGN eSUYAQ"
      data-testid="input-cpf"
      placeholder="Informe o seu CPF"
    />
  </div>
  <button
    class="sc-bdnylx kKDxdW"
    color="primary"
    data-testid="button-cpf"
    disabled=""
    type="button"
  >
    PROXIMO
  </button>
</DocumentFragment>
`;
